### PR TITLE
MM-37325: RHS update timer

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary

This PR adds the status update timer and the dropdown menu in the on-hover buttons. There is also some style polishing.

#### Videos

General walkthrough: a playbook with no default reminder is created (hence, the Post update button occupies the whole width at first). Then, an update is made with no reminder for the next one, so the button is in tertiary mode and the copy reads "Last update in...". Then an update is made with a reminder for the next one, so the button is in primary mode and the copy reads "Update due in...". Finally, the action of the three-dots button is demoed.

https://user-images.githubusercontent.com/3924815/127398181-4db7e087-3a11-4814-8328-bc0ed81beae3.mp4

Overdue updates and behaviour of the interactive post button Dismiss:

https://user-images.githubusercontent.com/3924815/127398247-99c105a2-1cb6-4d0d-b22c-c5c602dd4ffb.mp4

Overdue updates and behaviour of the interactive post button Update status:

https://user-images.githubusercontent.com/3924815/127398253-d7f9b072-31b6-499a-8438-e8cdbb5ce4ff.mp4

Quick view of the dark theme (I think the red here is a bit weird, but still readable):

https://user-images.githubusercontent.com/3924815/127398271-e4a6f090-b6a3-4354-8c65-0807c9070771.mp4

##### New run

| New run, no default reminder 	| New run, default reminder 	|
|---	|---	|
| image 	| image 	|

##### Scheduled
| Days 	| One hour 	| Minutes 	| 
|---	|---	|---	|
| image 	| image 	| image 	|

One minute 	| This minute 	| Now 	|
|---	|---	|---	|
 image 	| image 	| image 	|

##### Overdue
| This minute 	| One minute 	| Minutes 	| Hours 	|
|---	|---	|---	|---	|
| image 	| image 	| image 	| image 	|

##### No update scheduled

image

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-37325

#### Checklist

- [ ] Telemetry updated <- We did not talk about this in the meetings, so I'm just skipping and adding next week if needed
- [ ] ~~Gated by experimental feature flag~~
- [ ] ~~Unit tests updated~~

### CODE REVIEW

**Comment by agarciamontoro** [comment]
@justinegeffen, I need a review of the timer copy. This timer informs of when the next update is due (or when the last one was posted if no future one is scheduled) or how long has been overdue. A screenshot of what I'm exactly referring to:

image

One thing to take into account: the top line, which reads either "Last update", "Update due" or "Update overdue" is hardcoded text. The bottom line, with the time itself, comes from a component that handles the relatives times automatically, and I'm defaulting to what it says.


That said, the copy you need to review is the following list. I did not cover every possible case in the list, but the singular and plural works as expected, so this sample is enough:
- Last update:
  - `Last update this minute`
  - `Last update 1 minute ago`
  - `Last update 5 minutes ago`
  - `Last update 10 hours ago`
  - `Last update 30 days ago`
- Updates scheduled in the future:
  - `Update due in 30 days`
  - `Update due in 1 hour`
  - `Update due in 5 minutes`
  - `Update due in 1 minute`
  - `Update due this minute`
  - `Update due now`
- Updates missed:
  - `Update overdue this minute`
  - `Update overdue 1 minute ago`
  - `Update overdue 3 minutes ago`
  - `Update overdue 3 hours ago`
  - `Update overdue 30 days ago`
- Updates scheduled far in the future or missed looong ago. These ones will probably never happen, but they are technically possible:
  - `Update overdue February 03, 2020`
  - `Update due February 03, 2050`

**Comment by justinegeffen** [comment]
Thanks, @agarciamontoro! I think those suggestions all work perfectly. :)

**Review by justinegeffen — APPROVED** [review]

**Review by abhijit-singh — APPROVED** [review]
@agarciamontoro Perfect! 🎉

File: webapp/src/hooks/general.ts [hunk 1]
```diff
@@ -405,3 +406,20 @@ export function useFormattedUsernameByID(userId: string) {
 
     return useFormattedUsername(user);
 }
+
+export function useNow(refreshIntervalMillis = 1000) {
```
Comment (lieut-data) [inline webapp/src/hooks/general.ts:410]: Would it make sense to refresh every 5s vs 1s? I'm imagining the granularity here doesn't matter too much for correctness, but would also reduce rendering impact (plus unintentional side effects that might happen?).

**Review by lieut-data — APPROVED** [review]
Nice work, and excellent summary as always!

File: webapp/src/components/rhs/rhs_about_buttons.tsx [hunk 2]
```diff
@@ -80,8 +80,10 @@ const usePlaybookName = (playbookId: string) => {
 
     useEffect(() => {
         const getPlaybookName = async () => {
-            const playbook = await clientFetchPlaybook(playbookId);
-            setPlaybookName(playbook?.title || '');
+            if (playbookId !== '') {
+                const playbook = await clientFetchPlaybook(playbookId);
```
Comment (lieut-data) [inline webapp/src/components/rhs/rhs_about_buttons.tsx:84]: I think it's still feasible that we have an id but we can't find it (or access it?). Maybe we just `try / catch` here and handle the case?

Comment (agarciamontoro) [reply webapp/src/components/rhs/rhs_about_buttons.tsx:84]: Absolutely! Just wanted to quickly try if it fixes it (it does locally) while I implement an actual solution.

File: webapp/src/hooks/general.ts (hunk 1 above)
Comment (agarciamontoro) [reply webapp/src/hooks/general.ts:410]: Makes sense! Fixed in https://github.com/mattermost/mattermost-plugin-playbooks/pull/656/commits/acaaee3107a19a6926b872325002282d61492a37

**Review by lieut-data — APPROVED** [review]

File: webapp/src/components/duration.tsx [hunk 3]
```diff
@@ -30,19 +32,7 @@ export const renderDuration = (duration: moment.Duration) => {
...
-            setNow(moment());
-        };
-        const everySecond = 1000;
-        const timerId = setInterval(tick, everySecond);
-
-        return () => {
-            clearInterval(timerId);
-        };
-    }, []);
+    const now = useNow();
```
Comment (cpoile) [inline webapp/src/components/duration.tsx:35]: Nice refactor!

File: webapp/src/components/assets/icons/clock.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    width: 34px;
+    height: 34px;
+
+    color: var(--center-channel-color);
+`;
+
+const Clock = (props : {className?: string}) => (
+    <Svg
+        width='34'
+        height='34'
```
Comment (cpoile) [inline webapp/src/components/assets/icons/clock.tsx:19]: You can remove these (the Svg styled component already has them)

File: webapp/src/components/assets/icons/exclamation.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    width: 34px;
+    height: 34px;
+
+    color: var(--dnd-indicator);
+`;
+
+const Exclamation = (props : {className?: string}) => (
+    <Svg
+        width='38'
+        height='37'
```
Comment (cpoile) [inline webapp/src/components/assets/icons/exclamation.tsx:19]: same here

File: server/app/reminder.go [hunk 6]
```diff
@@ -158,3 +158,20 @@ func (s *PlaybookRunServiceImpl) removeReminderPost(playbookRunToModify *Playboo
...
+// ResetReminderTimer sets the previous reminder timer to 0.
+func (s *PlaybookRunServiceImpl) ResetReminderTimer(playbookRunID string) error {
+	playbookRunToModify, err := s.store.GetPlaybookRun(playbookRunID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve playbook run")
+	}
+
+	playbookRunToModify.PreviousReminder = 0
+	if err = s.store.UpdatePlaybookRun(playbookRunToModify); err != nil {
+		return errors.Wrapf(err, "failed to update playbook run after removing reminder post id")
```
Comment (cpoile) [inline server/app/reminder.go:171]: reminder post id?

File: webapp/src/components/dot_menu.tsx [hunk 7]
```diff
@@ -145,13 +145,14 @@ const DropdownMenuItemStyled = styled.a`
...
-export const DropdownMenuItem = (props: { text: string, onClick: () => void }) => {
+export const DropdownMenuItem = (props: { children: React.ReactNode, onClick: () => void, className?: string }) => {
     return (
         <DropdownMenuItemStyled
             href='#'
             onClick={props.onClick}
+            className={props.className}
         >
-            {props.text}
+            {props.children}
```
Comment (cpoile) [inline webapp/src/components/dot_menu.tsx:157]: oh yah, that's better

File: webapp/src/components/rhs/rhs_about_buttons.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        const getPlaybookName = async () => {
+            const playbook = await clientFetchPlaybook(playbookId);
+            setPlaybookName(playbook?.title || '');
+        };
+
+        getPlaybookName();
+    }, [playbookId]);
+
+    return playbookName;
+};
```
Comment (cpoile) [inline webapp/src/components/rhs/rhs_about_buttons.tsx]: I think we might need to make a ticket to start saving the playbooks in the store. This seems to be fetching the playbook from the server every time the playbookId changes. Not horrible, but it's definitely something we can clean up in the future.

**Review by cpoile — APPROVED** [review]
Awesome! Non-blocking comments. Thanks Alejandro!

File: webapp/src/components/duration.tsx (hunk 3 above)
Comment (agarciamontoro) [reply webapp/src/components/duration.tsx:35]: Thanks!

File: webapp/src/components/rhs/rhs_about_buttons.tsx (hunk 8 above)
Comment (agarciamontoro) [reply webapp/src/components/rhs/rhs_about_buttons.tsx]: Yeah, you're right! Ticket created: https://mattermost.atlassian.net/browse/MM-37514

File: webapp/src/components/assets/icons/clock.tsx (hunk 4 above)
Comment (agarciamontoro) [reply webapp/src/components/assets/icons/clock.tsx:19]: Addressed this in another PR (:man_facepalming:) https://github.com/mattermost/mattermost-plugin-playbooks/pull/658/commits/2bef5af3df9d089a1ae5758bf0284299a6477851

File: webapp/src/components/assets/icons/exclamation.tsx (hunk 5 above)
Comment (agarciamontoro) [reply webapp/src/components/assets/icons/exclamation.tsx:19]: Addressed this in another PR (:man_facepalming:) https://github.com/mattermost/mattermost-plugin-playbooks/pull/658/commits/2bef5af3df9d089a1ae5758bf0284299a6477851

File: server/app/reminder.go (hunk 6 above)
Comment (agarciamontoro) [reply server/app/reminder.go:171]: Addressed this in another PR (:man_facepalming:) https://github.com/mattermost/mattermost-plugin-playbooks/pull/658/commits/d128b3662aaa3695bc65d746469a71e237857535

**Pull request merged**